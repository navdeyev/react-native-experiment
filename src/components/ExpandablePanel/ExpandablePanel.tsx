import * as React from 'react';
import {Switch, Text, View} from 'react-native';
import styled from 'styled-components';

const Wrapper = styled(View)`
  border: 1px solid #333;
  margin-top: 5px;
`;

const PanelHeaderContainer = styled(View)`
  padding: 4px;
  display: flex;
  flex-direction: row;
  border: 0 solid #333;
  justify-content: space-between;
  border-bottom-width: ${ (props: State) => props.expanded ? '1px' : '0'};
`;

const PanelHeader = styled(View)`
  display: flex;
  justify-content: center;
`;

export const ChildrenContainer = styled(View)`
  padding: 4px;
`;

export interface Props {
    testID?: string;
    header: string;
    children: React.ReactNode
}

interface State {
    expanded: boolean
}

class ExpandablePanel extends React.Component<Props, State> {

    private static defaultProps = {
        testID: 'expandable-panel'
    };

    constructor(props: Props) {
        super(props);
        this.state = {expanded: false};
    }

    onValueChange = () => {
        this.setState((prevState: State) => {
            return {expanded: !prevState.expanded};
        });
    };

    render() {
        const {testID, header, children} = this.props;
        return (
            <Wrapper>
                <PanelHeaderContainer expanded={this.state.expanded}>
                    <PanelHeader><Text>{header}</Text></PanelHeader>
                    <Switch value={this.state.expanded}
                            onValueChange={this.onValueChange}
                            testID={`${testID}-switch`}/>
                </PanelHeaderContainer>
                {this.state.expanded ? <ChildrenContainer>{children}</ChildrenContainer> : null}
            </Wrapper>
        );
    }

}

export default ExpandablePanel;
