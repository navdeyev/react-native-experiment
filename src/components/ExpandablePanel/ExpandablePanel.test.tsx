import * as React from 'react';
import {shallow} from 'enzyme';
import {Text} from 'react-native';
import ExpandablePanel, {ChildrenContainer, Props} from './ExpandablePanel';

describe('ExpandablePanel', () => {

    let props: Props;
    beforeEach(() => {
        props = {
            children: <Text>A journey of a thousand miles starts with a single step</Text>,
            header: 'View a quote of the day'
        }
    });

    it('does not display children in initial state', () => {
        const rendered = shallow(<ExpandablePanel {...props} />);
        expect(rendered.find(ChildrenContainer).exists()).toBe(false);
        expect(rendered.getElement()).toMatchSnapshot();
    });

    it('displays children in expanded state', () => {
        const rendered = shallow(<ExpandablePanel {...props} />);
        const switchControl = rendered.find('[testID="expandable-panel-switch"]');
        switchControl.simulate('valueChange');
        expect(rendered.find(ChildrenContainer).exists()).toBe(true);
        expect(rendered.getElement()).toMatchSnapshot();
    });
});
