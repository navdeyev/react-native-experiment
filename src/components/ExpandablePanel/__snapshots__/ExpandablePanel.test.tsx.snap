// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpandablePanel displays children in expanded state 1`] = `
<Styled(Component)>
  <Styled(Component)
    expanded={true}
  >
    <Styled(Component)>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        View a quote of the day
      </Text>
    </Styled(Component)>
    <Switch
      disabled={false}
      onValueChange={[Function]}
      testID="expandable-panel-switch"
      value={true}
    />
  </Styled(Component)>
  <Styled(Component)>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      A journey of a thousand miles starts with a single step
    </Text>
  </Styled(Component)>
</Styled(Component)>
`;

exports[`ExpandablePanel does not display children in initial state 1`] = `
<Styled(Component)>
  <Styled(Component)
    expanded={false}
  >
    <Styled(Component)>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        View a quote of the day
      </Text>
    </Styled(Component)>
    <Switch
      disabled={false}
      onValueChange={[Function]}
      testID="expandable-panel-switch"
      value={false}
    />
  </Styled(Component)>
</Styled(Component)>
`;
